purple-pill/js/controllers/fontexportController.js

```diff
@@ -20,152 +92,136 @@ function($scope, $http, sharedScope, $timeout) {
 
         function setDownloadBlobLink(text, blob, filename) {
-            $("#progressbar").animate({"width": "100%", "opacity": 1}, /*duration:*/ UI_UPDATE_TIMESLICE);
-            $("#progresslabel").html("");
-            $("#blob_download").css("display", "block");
-            $("#blob_download").children("a").html(text).click(function(){
+            var download = $("#blob_download");
+            progress.complete();
+
+            download.css("display", "block");
+            download.children("a").html(text).click(function(){
                 $scope.data.stateless.saveAs(blob, filename);
-                resetProgressBar();
-                delete bundle_data;
+                progress.reset();
+                download.css("display", "none").children("a").unbind("click");
             });
         }
-        
-        function resetProgressBar() {
-            $("#progressbar").animate({"opacity": 0, "width": 0}, /*duration:*/ 0); // (that means "do it immediately!")
-            $("#progresslabel").html("");
-            $("#blob_download").css("display", "none").children("a").unbind("click");
-        }
 
-        function exportFont_compute_glyphs(){
-            var text
-              , current_glyph
-              , total_glyphs
-              , glyph_name
-              , it = generator.next()
-              ;
-            if (!it.done){
-                current_glyph = it.value['current_glyph'];
-                total_glyphs = it.value['total_glyphs'];
-                glyph_id = it.value['glyph_id'];
-                text = total_instances + " instances to export. Calculating glyph '" + glyph_id + "' (" + (current_glyph+1) + " of " + total_glyphs + ") of instance #" + (current_instance+1);
-                setProgress(glyphs_phase_percentage * (current_instance + ((current_glyph+1) / total_glyphs))/total_instances, text);
-                $timeout(exportFont_compute_glyphs, UI_UPDATE_TIMESLICE);
-            } else {
-                delete generator;
-                if (++current_instance < total_instances) {
-                    generator = getExportGenerator(current_instance);
-                    $timeout(exportFont_compute_glyphs, UI_UPDATE_TIMESLICE);
-                } else {
-                    current_instance = 0;
-                    text = total_instances + " instances to export in .ufo.zip format, zipping instance #" + (current_instance+1) + " (can take a while)";
-                    setProgress(glyphs_phase_percentage, text);
-                    $timeout(exportFont_pack_instance_chunk, UI_UPDATE_TIMESLICE);
+        // note how the first three args are bound, new Promise will call the bound function with
+        // the two missing args `resolve` and `reject`
+        return new Promise(
+            _exportFontComputeGlyphs.bind(null, exportObjects, totalInstances, bundleFolder)
+        ).then(function(){
+            setDownloadBlobLink(bundleFileName, bundle.generate({type:"blob"}), bundleFileName);
```

Comment:
if you do here just something like:
```
return [bundleFileName, bundle.generate({type:"blob"})];
```

and in the caller:

```js
exportIsRunning = true;
exportFont_compute_glyphs(exportObjects).then(function(result) {
    setDownloadBlobLink(bundleFileName, result[1], result[0]);
    exportIsRunning = false;
});
```

Then you have decoupled the  export routine from it's environment. That could be helpful if we can chose one day for example to save the result directly to disk. A function called `exportFont_compute_glyphs` does not need to know how to make a download link.